open-craft/learner-profile-dashboard
# Send learner profile data to adaptive engine.

cf. [OC-4387](https://tasks.opencraft.com/browse/OC-4387)

This PR adds logic for sending learner profile data (group membership information and scores from quantitative questions) to VPAL's adaptive engine.

It also re-enables pylint's `missing-docstring` check and ships all docstrings that were missing up until now.

**Test instructions**

*Note that we can't test the happy path manually at this stage because the endpoint to which the LPD sends learner profile data has not been added to the [adaptive engine](https://github.com/harvard-vpal/adaptive-engine) yet; and adding this endpoint to the adaptive engine is out of scope for us; VPAL is taking care of that part of the integration. However, automated tests that ship with this PR build confidence that the happy path is working as expected.*

1. Set up an LPD with both qualitative and quantitative questions, and make sure that the LPD can process qualitative data (i.e., that you have `GROUP_KCS` and corresponding knowledge components set up correctly). If necessary, use instructions and data from previous PRs to help you along.
1. Add new settings to `app/local_settings.py`:
    ```python
    # Domain of the Open edX instance that this LPD deployment is connected to
    OPENEDX_INSTANCE_DOMAIN = 'development.domain.com'
    # URL of the Adaptive Engine deployment that this LPD deployment is connected to
    ADAPTIVE_ENGINE_URL = 'http://localhost'
    # Auth token for requests to Adaptive Engine deployment that this LPD deployment is connected to
    ADAPTIVE_ENGINE_TOKEN = 'development-token'
    ```
1. Temporarily make
    ```python
     print(user_id)
     ```
     the first line of the body of the `authentication_hook` method in `app/__init__.py`.
1. Temporarily insert
    ```python
    print(payload)
    ```
    before the last line of `lpd/client.py`. This will allow you to inspect the data that the LPD sends to the adaptive engine later on.
1. Access the LPD in Studio/LMS, via a unit that contains it. Check the development server's output for the LTI user ID of the current user, and make a note of it. (If the server doesn't output an LTI user ID, simply navigate to the LPD in LMS/Studio. That is, if you accessed the LPD via Studio first, access it again via the LMS, or vice versa. Explanation: The `authentication_hook` only runs when encountering a different LTI user ID, and Studio and LMS report different LTI user IDs for the same user. This means that you can force `authentication_hook` to run by loading the LPD in the Open edX component that you didn't use on the first try.)
1. Open the browser console.
1. Provide answers to LPD questions and click "Submit your answers". Observe that the browser console reports a 500 error with a message saying
    > Could not transmit scores to adaptive engine.
1. Check the development server's output for the contents of the payload that the LPD tried to send to the adaptive engine. Make sure that:
    * `lti_user_id` should match the LTI user ID that you wrote down while following the steps above. It should be different from the `username` of the corresponding `User` that is stored in the LPD's database.
    * `tool_consumer_instance_guid` matches `OPENEDX_INSTANCE_DOMAIN`.
    * `masteries` contains entries for:
        * all knowledge components representing groups (i.e., all `GROUP_KCS`)
        * all answer options that belong to questions from the current section and that are configured to `influence_recommendations`.

**Reviewers**
